adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -0,0 +1,400 @@
+    }
+
+    /**
+     * Creates the router for handling requests.
+     * 
+     * @return The router.
+     */
+    private Router createRouter() {
+        final Router router = Router.router(vertx);
+        LOG.info("limiting size of inbound request body to {} bytes", config.getMaxPayloadSize());
+        router.route().handler(BodyHandler.create().setBodyLimit(config.getMaxPayloadSize()));
+
+        String statusResourcePath = getStatusResourcePath();
+        if (statusResourcePath != null) {
+            router.route(HttpMethod.GET, statusResourcePath).handler(this::doGetStatus);
+        }
+
+        addRoutes(router);
+        return router;
+    }
+
+    /**
+     * Returns the path for the status resource. By default, this is "/status".
+     * Subclasses may override this method to return a different path or to return null, 
+     * in which case the status resource will be disabled.
+     * 
+     * @return path or null
+     */
+    protected String getStatusResourcePath() {
+        return "/status";
+    }
+
+    /**
+     * Adds the routes for handling requests to this adapter.
+     * 
+     * The handler of each route is to invoke one of the uploadTelemetryMessage() 
+     * or uploadEventMessage() methods.
+     * 
+     * @param router router to add the routes to
+     */
+    protected abstract void addRoutes(Router router);
+
+    private void bindHttpServer(final Router router, final Future<Void> startFuture) {
+        HttpServerOptions options = new HttpServerOptions();
+        options.setHost(config.getBindAddress()).setPort(config.getPort()).setMaxChunkSize(4096);
+        server = vertx.createHttpServer(options);
+        server.requestHandler(router::accept).listen(done -> {
+            if (done.succeeded()) {
+                LOG.info("adapter running on {}:{}", config.getBindAddress(), server.actualPort());
+                startFuture.complete();
+            } else {
+                LOG.error("error while starting up adapter", done.cause());
+                startFuture.fail(done.cause());
+            }
+        });
+    }
+
+    @Override
+    public final void stop(Future<Void> stopFuture) throws Exception {
+        Future<Void> shutdownTracker = Future.future();
+        shutdownTracker.setHandler(done -> {
+            if (done.succeeded()) {
+                LOG.info("adapter has been shut down successfully");
+                stopFuture.complete();
+            } else {
+                LOG.info("error while shutting down adapter", done.cause());
+                stopFuture.fail(done.cause());
+            }
+        });
+
+        Future<Void> serverTracker = Future.future();
+        if (server != null) {
+            server.close(serverTracker.completer());
+        } else {
+            serverTracker.complete();
+        }
+        Future<Void> afterServerStopTracker = Future.future();
+        serverTracker.compose(d -> {
+            afterHttpServerStop(afterServerStopTracker);
```

Comment:
Is there a reason why you want to invoke the subclass housekeeping *after* you have stopped the server? I noticed the same with the startup process, where you invoke the subclass method *before* starting up the http server. IMHO this should be exactly the other way around, i.e. start up the server, *then* invoke subclass and first invoke subclass, *then* shut down http server ...